stream: pipeline wait for close before calling the callback

jakecastelli opened the pull request:
> // Use the pipeline API to easily pipe a series of streams
// together and get notified when the pipeline is fully done.

> stream.pipeline() closes all the streams when an error is raised

pipeline doc: [here](https://nodejs.org/api/stream.html#streampipelinesource-transforms-destination-callback)

The pipeline should wait for close event to finish before calling the callback as that should be the proper "fully done" state.

The `finishCount` should not below 0 when calling `finishImpl` function (should not be negative).

Fixes: https://github.com/nodejs/node/issues/51540

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

jakecastelli commented:
I have spent a lot of time looking into this issue, I would really appreciate your comment or feedback 🙏 
@wh0 @lpinca @ronag

wh0 commented:
This is different from what I was picturing, and I wonder if it works with a pipeline that has some number of `PassThrough`s in between.

The problem as I see it is that over time, various new places have been added that call `finish` that circumstantially deliver specific error conditions and which don't fit into the `finishCount` logic. I see the fix being along the lines of this:

```diff
diff --git a/lib/internal/streams/pipeline.js b/lib/internal/streams/pipeline.js
index bb34759..83c53d8 100644
--- a/lib/internal/streams/pipeline.js
+++ b/lib/internal/streams/pipeline.js
@@ -225,6 +225,10 @@ function pipelineImpl(streams, callback, opts) {
     finishImpl(err, --finishCount === 0);
   }
 
+  function finishOnlyHandleError(err) {
+    finishImpl(err, false);
+  }
+
   function finishImpl(err, final) {
     if (err && (!error || error.code === 'ERR_STREAM_PREMATURE_CLOSE')) {
       error = err;
@@ -279,7 +283,7 @@ function pipelineImpl(streams, callback, opts) {
           err.name !== 'AbortError' &&
           err.code !== 'ERR_STREAM_PREMATURE_CLOSE'
         ) {
-          finish(err);
+          finishOnlyHandleError(err);
         }
       }
       stream.on('error', onError);
@@ -372,7 +376,7 @@ function pipelineImpl(streams, callback, opts) {
     } else if (isNodeStream(stream)) {
       if (isReadableNodeStream(ret)) {
         finishCount += 2;
-        const cleanup = pipe(ret, stream, finish, { end });
+        const cleanup = pipe(ret, stream, finish, finishOnlyHandleError, { end });
         if (isReadable(stream) && isLastStream) {
           lastStreamCleanup.push(cleanup);
         }
@@ -415,12 +419,12 @@ function pipelineImpl(streams, callback, opts) {
   return ret;
 }
 
-function pipe(src, dst, finish, { end }) {
+function pipe(src, dst, finish, finishOnlyHandleError, { end }) {
   let ended = false;
   dst.on('close', () => {
     if (!ended) {
       // Finish if the destination closes before the source has completed.
-      finish(new ERR_STREAM_PREMATURE_CLOSE());
+      finishOnlyHandleError(new ERR_STREAM_PREMATURE_CLOSE());
     }
   });
```

(not tested)

mcollina reviewed (approved):
lgtm

jakecastelli commented:
@wh0 Thanks for spending the time and taking the time look into it!

> I wonder if it works with a pipeline that has some number of PassThroughs in between

That was a great catch! It WON'T work when I add one or more `PassThrough`.

> The problem as I see it is that over time, various new places have been added that call finish that circumstantially deliver specific error conditions and which don't fit into the finishCount logic. I see the fix being along the lines of this

I see your point and I agree, overtime it would become harder to work with. And I have also tested it your patch (a minor syntax error fix for pipe - as its positional args, no biggie) and it works just like expected.

I will temporarily mark this PR as draft and take some more time to look into your patch and add additional tests to cover the case where there one or more `PassTrough`.

@mcollina Thanks for looking into it, will bug you for review again once its ready 😄 🙏

jakecastelli commented:
I have added another new test to include a number of `PassThrough` in between.

@wh0 your fix was clean and way better than my original "monkey patch" fix (there was no issue with your patch, my previous message saying there was a little bit issue was just me waking up without coffee 😄) absolutely learned a lot. I have added you as a co-author please let me know if you are happy with me using your patch.

ronag commented:
I would like to properly review this. Please don't land yet.

wh0 commented:
thanks jakecastelli, glad to know this approach works even with the PassThroughs. ronag thanks for your time as well

mcollina reviewed (approved):
lgtm

mcollina commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/3447/

ronag reviewed (changes requested):
I believe this will break legacy user land streams that don't always emit close. Pipeline does not guarantee that close is emitted on all streams before the callback is invoked. The current behavior is intentional. If the docs says otherwise I believe we should fix those instead.

This was mentioned in the comments for the issue: https://github.com/nodejs/node/issues/51540

jakecastelli commented:
Thanks @ronag! This makes sense, when callback is not called in the `destroy`, then the callback would not be called in the pipeline either.Even though this behaviour feels slightly weird as if there is any

jakecastelli commented:
Thanks @ronag! This makes sense, when callback is not called in the `destroy`, then the callback would not be called in the `pipeline` either.

Even though the current behaviour feels slightly weird as if there is any async operation in the `destroy` such as cleanup, the user cannot rely on the callback that cleanup is guaranteed done, I believe this patch would introduce more potential issues than its fixing.

Thanks again for everyone's time on this, I will leave it open for another day and then close this PR 🙏

wh0 commented:
> I believe this will break legacy user land streams that don't always emit close.

what kind of break is this referring to? e.g. never calling the callback?

wh0 commented:
the patch doesn't depend on `close` being emitted. the two `finalCount` decrements come from `eos(src)` and `eos(dst)`. end-of-stream has its own tests that ensure it works properly on `emitClose: false` streams

ronag commented:
I will take another look.

jakecastelli commented:
> e.g. never calling the callback?

Never calling the callback in `destroy` would cause `callback` in the pipeline not running - I think that's one big user land breaking change?

Another one I found is when you use the `pipeline` with `PassThrough` from `node:stream/promises` version it would give you warning about `MaxListenersExceededWarning`.

```
(node:29863) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [PassThrough]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
```

I haven't dug into the second one yet, will take a look later.

mcollina commented:
@ronag I think that case is handled.

ronag reviewed (approved):
LGTM. Excellent work.

jakecastelli commented:
Green CI 🎉 

Before this land, I wanted to ask about this:

> when callback is not called in the destroy, then the callback would not be called in the pipeline either

Is this a breaking change?

wh0 commented:
I think this behavior changed last between 17.2 and 17.3

mcollina commented:
@jakecastelli why is it breaking? @wh0 have you got a reference?

mcollina commented:
I've added a "backing-for-lts" label, so we don't backport to v20 for a bit.

mcollina commented:
Unless we have an example of something that regressed, I think this can land.

ronag commented:
I don't think is breaking.

jakecastelli commented:
> I think this behavior changed last between 17.2 and 17.3

I tested it on 16.x, 17.2, 17.3, 17.x didn't see the behaviour change though

> @jakecastelli why is it breaking?

With this patch change when the callback is not called in the `destroy` method, the call won't be called in the `pipeline`, to me this sounds like a breaking change (even though this should be the "right" behaviour?). See the example code below for example:

```js
const stream = require("stream");

const src = new stream.Readable();
const dst = new stream.Writable({
  write(chunk, encoding, callback) {
    callback();
  },
  destroy(error, callback) {
    //   takes a while to destroy
     setImmediate(() => {
    // not calling callback here will result callback not be called in the stream.pipeline
    //   callback(error);
     });
  },
});

stream.pipeline(src, dst, (err) => {
  console.log(`pipeline done, err=${err.message}, dst.closed=${dst.closed}`);
});
src.destroy(new Error("problem"));
```

ronag commented:
> With this patch change - when the callback is not called in the destroy method, the callback won't be called in the pipeline, to me this sounds like a breaking change (even though this should be the "right" behaviour?). See the example code below for example:

You always have to call the callback. If you don't it's broken in so many way anyway that I wouldn't consider it a breaking change.

jakecastelli commented:
> You always have to call the callback. If you don't it's broken in so many way anyway that I wouldn't consider it a breaking change.

Ah I see, thanks for confirming this!

jakecastelli commented:
I think we are ready to land @mcollina 👍

ljharb commented:
Any chance this is related to https://github.com/npm/cli/issues/7657 ?

jakecastelli commented:
Hi @ljharb based on the discussion in 53904 and a few others I think it is not related to this one. I left my finding in https://github.com/npm/cli/issues/7657#issuecomment-2237071834 about `v22.5.0` on arm64 build though.
